Add section on dbt provider to dbt guide (#361)
astronomer/airflow-guides: guides/airflow-dbt.md

## Overview

[dbt](https://getdbt.com/) is an open-source library for analytics engineering that helps users build interdependent SQL models for in-warehouse data transformation. As ephemeral compute becomes more readily available in data warehouses thanks to tools like [Snowflake](https://snowflake.com/), dbt has become a key component of the modern data engineering workflow. Now, data engineers can use dbt to write, organize, and run in-warehouse transformations of raw data.

Teams can use Airflow to orchestrate and execute dbt models as DAGs. Running dbt with Airflow ensures a reliable, scalable environment for models, as well as the ability to trigger models only after every prerequisite task is met. Airflow also gives you fine-grained control over dbt tasks such that teams have observability over every step in their dbt models.

In this guide, we'll walk through:

> - Using the [dbt Provider](https://registry.astronomer.io/providers/dbt-cloud) to orchestrate dbt Cloud with Airflow.

Reviewer: ```suggestion
- Using the [dbt Cloud Provider](https://registry.astronomer.io/providers/dbt-cloud) to orchestrate dbt Cloud with Airflow.
```

- Two common use cases for orchestrating dbt Core with Airflow via the `BashOperator`: 
    - At the [project](https://docs.getdbt.com/docs/building-a-dbt-project/projects) level and, 
    - At the [model](https://docs.getdbt.com/docs/building-a-dbt-project/building-models) level. 
- How to extend the model-level use case by automating changes to a dbt model.

> The dbt Core sections of this guide summarize some of the key practices and findings from our blog series with [Updater](https://updater.com/) and [Sam Bail](https://sambail.com/) about using dbt in Airflow. For more information, check out [Part 1](https://www.astronomer.io/blog/airflow-dbt-1), [Part 2](https://www.astronomer.io/blog/airflow-dbt-2), and [Part 3](https://www.astronomer.io/blog/airflow-dbt-3) of the series.

## dbt Cloud

> To orchestrate [dbt Cloud](https://www.getdbt.com/product/what-is-dbt/) jobs with Airflow, you can use the [dbt Provider](https://registry.astronomer.io/providers/dbt-cloud), which contains the following useful modules:

Reviewer: ```suggestion
To orchestrate [dbt Cloud](https://www.getdbt.com/product/what-is-dbt/) jobs with Airflow, you can use the [dbt Cloud Provider](https://registry.astronomer.io/providers/dbt-cloud), which contains the following useful modules:
```

- **`DbtCloudRunJobOperator`:** Executes a dbt Cloud job.
- **`DbtCloudGetJonRunArtifactOperator`:** Downloads artifacts from a dbt Cloud job run.
- **`DbtCloudJobRunSensor`:** Waits for a dbt Cloud job run to complete.

> In order to use the dbt Provider in your DAGs, you will need to complete the following steps:

Reviewer: ```suggestion
In order to use the dbt Cloud Provider in your DAGs, you will need to complete the following steps:
```

1. Add the `apache-airflow-providers-dbt-cloud` package to your Airflow environment. If you are working in an Astro project, you can add the package to your `requirements.txt` file.

> 2. Set up an Airflow connection to your dbt Cloud instance. The connection type should be `dbt Cloud`, and it should include an API token from your dbt Cloud account. If you want your dbt Provider tasks to use a default account ID, you can add that to the connection, but it is not required.

Reviewer: ```suggestion
2. Set up an Airflow connection to your dbt Cloud instance. The connection type should be `dbt Cloud`, and it should include an API token from your dbt Cloud account. If you want your dbt Cloud Provider tasks to use a default account ID, you can add that to the connection, but it is not required.
```

Reviewer: Should also be a link to the [connection doc](https://airflow.apache.org/docs/apache-airflow-providers-dbt-cloud/stable/connections.html) in Airflow here?

Author: Sure, can add one

> In the DAG below, we show a simple implementation of the dbt Provider. This example showcases how to run a dbt Cloud job from Airflow, while adding an operational check to ensure the dbt Cloud job is not running prior to triggering. The `DbtCloudHook` provides a `list_job_runs()` method which can be used to retrieve the latest triggered run for a job and check the status of said run. If the job is not in a state of 10 (Success), 20 (Error), or 30 (Cancelled), the pipeline will not try to trigger it.

Reviewer: ```suggestion
In the DAG below, we show a simple implementation of the dbt Cloud Provider. This example showcases how to run a dbt Cloud job from Airflow, while adding an operational check to ensure the dbt Cloud job is not running prior to triggering. The `DbtCloudHook` provides a `list_job_runs()` method which can be used to retrieve all runs for a given job. The operational check uses this method to retrieve the latest triggered run for a job and check its status. If the job is currenly not in a state of 10 (Success), 20 (Error), or 30 (Cancelled), the pipeline will not try to trigger another run.
```

Reviewer: >If the job is not in a state of 10 (Success), 20 (Error), or 30 (Cancelled), the pipeline will not try to trigger it.

This unfortunately will all change when V4 of the dbt Cloud API is released and the provider is updated. Allegedly happening some time this year. Just a preemptive warning.

Reviewer: Should the `DbtCloudHook` be listed as a "useful module" as well above?

def _check_job_not_running(job_id):
    """
    Retrieves the last run for a given dbt Cloud job and checks to see if the job is not currently running.
    """
    hook = DbtCloudHook(DBT_CLOUD_CONN_ID)
    runs = hook.list_job_runs(job_definition_id=job_id, order_by="-id")
    latest_run = runs[0].json()["data"][0]

    return DbtCloudJobRunStatus.is_terminal(latest_run["status"])

@dag(
    start_date=datetime(2022, 2, 10),
    schedule_interval="@daily",
    catchup=False,
    default_view="graph",
    doc_md=__doc__,
)
def check_before_running_dbt_cloud_job():
    begin, end = [DummyOperator(task_id=id) for id in ["begin", "end"]]

    check_job = ShortCircuitOperator(
        task_id="check_job_is_not_running",
        python_callable=_check_job_not_running,
        op_kwargs={"job_id": JOB_ID},
    )

    trigger_job = DbtCloudRunJobOperator(
        task_id="trigger_dbt_cloud_job",
        dbt_cloud_conn_id=DBT_CLOUD_CONN_ID,
        job_id=JOB_ID,
        check_interval=600,
        timeout=3600,
    )

    begin >> check_job >> Label("Job not currently running. Proceeding.") >> trigger_job >> end

dag = check_before_running_dbt_cloud_job()
```

> Note that in the `DbtCloudRunJobOperator` you must provide the dbt connection ID as well as the `job_id` of the job you are triggering.

Reviewer: Technically dbt Cloud connection ID doesn't need to passed in directly if folks use the default conn ID, but maybe this is not a specific that we typically caveat in the guides?

Author: Yeah I think better to explicitly show it being passed in so people realize a connection is needed. Won't hurt anything if they pass the default in, even if it's not necessary.

> > **Note:** The full code for this example, along with other DAGs that implement the dbt Provider, can be found on the [Astronomer Registry](https://registry.astronomer.io/dags?providers=dbt+Cloud&page=1).

Reviewer: ```suggestion
> **Note:** The full code for this example, along with other DAGs that implement the dbt Cloud Provider, can be found on the [Astronomer Registry](https://registry.astronomer.io/dags?providers=dbt+Cloud&page=1).
```

## dbt Core

> When orchestrating dbt Core with Airflow, the most straight forward DAG design is to run dbt commands directly through the [`BashOperator`](https://airflow.apache.org/docs/apache-airflow/stable/howto/operator/bash.html). In this section we'll show a couple examples for how to do so.

Reviewer: >... the most straight forward DAG design

Maybe phrase it as "A straightforward DAG design"?

Reviewer: Should we link to the `BashOperator` page in the Registry instead?

Author: Yes, good call

@dag(
    start_date=datetime(2022, 2, 10),
    schedule_interval="@daily",
    catchup=False,
    default_view="graph",
    doc_md=__doc__,
)
def check_before_running_dbt_cloud_job():
    begin, end = [DummyOperator(task_id=id) for id in ["begin", "end"]]

    check_job = ShortCircuitOperator(
        task_id="check_job_is_not_running",
        python_callable=_check_job_not_running,
        op_kwargs={"job_id": JOB_ID},
    )

    trigger_job = DbtCloudRunJobOperator(
        task_id="trigger_dbt_cloud_job",
        dbt_cloud_conn_id=DBT_CLOUD_CONN_ID,
        job_id=JOB_ID,
        check_interval=600,
        timeout=3600,
    )

    begin >> check_job >> Label("Job not currently running. Proceeding.") >> trigger_job >> end

dag = check_before_running_dbt_cloud_job()
```

Note that in the `DbtCloudRunJobOperator` you must provide the dbt connection ID as well as the `job_id` of the job you are triggering.

> **Note:** The full code for this example, along with other DAGs that implement the dbt Cloud Provider, can be found on the [Astronomer Registry](https://registry.astronomer.io/dags?providers=dbt+Cloud&page=1).

## dbt Core

> When orchestrating dbt Core with Airflow, a straight forward DAG design is to run dbt commands directly through the [`BashOperator`](https://registry.astronomer.io/providers/apache-airflow/modules/bashoperator). In this section we'll show a couple examples for how to do so.

Reviewer: ```suggestion
When orchestrating dbt Core with Airflow, a straightforward DAG design is to run dbt commands directly through the [`BashOperator`](https://registry.astronomer.io/providers/apache-airflow/modules/bashoperator). In this section we'll show a couple examples for how to do so.
```

## Conclusion

> To recap, in this guide we have learned about dbt, how to create dbt tasks in Airflow, and how to productionize those tasks to automatically create tasks based on a manifest. For a more detailed discussion on trade-offs, limitations, and adding dbt Core to a full ELT pipeline, see our blog posts. To see more examples of how to use dbt and Airflow to build pipelines, check out our [dbt DAGs on the Registry](https://registry.astronomer.io/dags/?query=dbt&badges=certified).

Reviewer: ```suggestion
To recap, in this guide we have learned about dbt Cloud and dbt Core, how to create and productionize dbt tasks in Airflow, and how to automatically create dbt Core tasks based on a manifest. For a more detailed discussion on trade-offs, limitations, and adding dbt Core or dbt Cloud to a full ELT pipeline, see our blog posts. To see more examples of how to use dbt and Airflow to build pipelines, check out our [dbt DAGs on the Registry](https://registry.astronomer.io/dags/?query=dbt&badges=certified).
```
